### PR TITLE
BUG: support min/max functions for rolling windows with custom BaseIndexer

### PR DESCRIPTION
**Pull request by AlexKirko**
- [X] xref #32865 
- [X] 1 tests added / 1 passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

## The problem
We currently don't support several rolling window functions when building a rolling window object using a custom class descended from `pandas.api.indexers.Baseindexer`. The implementations were written with backward-looking windows in mind, and this led to these functions breaking.
Currently, using these functions returns a `NotImplemented` error thanks to #33057, but ideally we want to update the implementations, so that they will work without a performance hit. This is what I aim to do over a series of PRs.

## Scope of PR
This PR proposes an alternative implementation for the `min` and `max` rolling window functions implemented in the `_roll_min_max_variable` function in the `aggregations.pyx` file.

## Perf note
This implementation is slightly faster than the one on master branch (tested with a `Series` with 10e6 randomly generated values and `timeit`). If necessary, I can add a benchmark to the benchmark suite, although it must be noted that we need to benchmark with large arrays of data to minimize the effect of overhead.

### CODE REVIEW

**Comment by AlexKirko** [comment]
The core idea is that we ancor the output index we write at to the index in the value array it corresponds to. This generalizes the function and also removes the need to fiddle with `closed_offset`.

**Comment by AlexKirko** [comment]
On another note, with these changes, it is no longer necessary to have separate code for initializing the first value and for writing the final value. Everything can be done in the main loop.

File: pandas/tests/window/test_rolling.py [hunk 1]
```diff
@@ -465,3 +466,35 @@ def test_rolling_count_default_min_periods_with_null_values(constructor):
...
+                    start[i] = i
+                    end[i] = min(i + self.window_size, num_values)
+                else:
+                    start[i] = i
+                    end[i] = i + 1
+            return start, end
+
+    values = np.arange(10)
+
+    indexer = ForwardIndexer(window_size=3)
```
Comment (jreback) [inline pandas/tests/window/test_rolling.py]: hmm are you anticpating exposing this to users? should we be using this internally?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/window/test_rolling.py (hunk 1 above)
Comment (AlexKirko) [reply pandas/tests/window/test_rolling.py]: Do you mean exposing a forward-looking rolling window as part of our default API?
I don't think that is necessary: my goal is to make sure that our implementations are robust against custom indexers with arbitrary offsets relative to the position in the input array they are calculated for.
The `ForwardIndexer` class is used in testing for convenience as it is, by far, the most common customization people expect us to support, in my limited experience.

File: pandas/tests/window/test_rolling.py [hunk 2]
```diff
@@ -465,3 +466,36 @@ def test_rolling_count_default_min_periods_with_null_values(constructor):
...
+
+@pytest.mark.parametrize("constructor", [Series, DataFrame])
+@pytest.mark.parametrize(
+    "func,expected",
+    [
+        ("min", [0.0, 1.0, 2.0, 3.0, 4.0, 6.0, 6.0, 7.0, 8.0, np.nan]),
+        ("max", [2.0, 3.0, 4.0, 100.0, 100.0, 100.0, 8.0, 9.0, 9.0, np.nan]),
+    ],
+)
+def test_rolling_forward_window(constructor, func, expected):
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py]: Could you put this test in `pandas/tests/window/test_base_indexer.py`?

File: pandas/tests/window/test_rolling.py [hunk 3]
```diff
@@ -465,3 +466,36 @@ def test_rolling_count_default_min_periods_with_null_values(constructor):
...
+                    start[i] = i
+                    end[i] = i + 1
+            return start, end
+
+    values = np.arange(10)
+    values[5] = 100.0
+
+    indexer = ForwardIndexer(window_size=3)
+    rolling = constructor(values).rolling(window=indexer, min_periods=2)
+    result = getattr(rolling, func)()
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py]: Can we also test this against `rolling.apply(lambda x: max(x))` (or the equivalent)?

File: pandas/tests/window/test_rolling.py (hunk 2 above)
Comment (AlexKirko) [reply pandas/tests/window/test_rolling.py]: Done.

File: pandas/tests/window/test_rolling.py (hunk 3 above)
Comment (AlexKirko) [reply pandas/tests/window/test_rolling.py]: Done.

Comment (AlexKirko) [reply pandas/tests/window/test_rolling.py]: @mroeschke 
Should we drop testing against an explicitly provided array then? Doin both seems redundant to me.

Comment (mroeschke) [reply pandas/tests/window/test_rolling.py]: I think it's okay to keep the explicit array as the source of truth on the off chance that `rolling(Indexer).max()` and `rolling.apply(lambda x: max(x)` both incorrectly change in the same direction

Comment (AlexKirko) [reply pandas/tests/window/test_rolling.py]: Okay, let's keep it then. I was afraid of something like this, which is why I started by testing against an explicit array and not against apply.

File: pandas/tests/window/test_rolling.py (hunk 1 above)
Comment (mroeschke) [reply pandas/tests/window/test_rolling.py]: Might be interesting to provide a standard set of `BaseIndexer`s in the library. `ForwardIndexer` may be one that we provide once we standardize the definition and API.

@AlexKirko feel free to open a new issue about including `ForwardIndexer` and `BackwardIndexer` in the `pandas.api.indexers`

File: pandas/_libs/window/aggregations.pyx [hunk 4]
```diff
@@ -1068,60 +1068,45 @@ cdef _roll_min_max_variable(ndarray[numeric] values,
...
-            W.push_back(i)
-
-        # if right is open then the first window is empty
-        close_offset = 0 if endi[0] > starti[0] else 1
         # first window's size
         curr_win_size = endi[0] - starti[0]
+        # GH 32865
+        # Anchor output index to values index to provide custom
+        # BaseIndexer support
+        for i in range(N):
```
Comment (mroeschke) [inline pandas/_libs/window/aggregations.pyx]: Mind running the ASVs for rolling max/min to see if the existing functions don't suffer a performance hit?

Comment (AlexKirko) [reply pandas/_libs/window/aggregations.pyx:1076]: @mroeschke
Sure! Sorry for missing that we have a benchmark for this stuff in the suite.

Ran the benchmarks in `rolling.Methods`. No significant changes.

Detailed output is below.

<details>
```
[100.00%] ··· rolling.Methods.time_rolling                                                                            ok
[100.00%] ··· ============= ======== ======= ======== =============
               constructor   window   dtype   method
              ------------- -------- ------- -------- -------------
                DataFrame      10      int    median     74.0±1ms
                DataFrame      10      int     mean     13.7±0.1ms
                DataFrame      10      int     max      9.12±0.1ms
                DataFrame      10      int     min      9.09±0.2ms
                DataFrame      10      int     std      16.2±0.2ms
                DataFrame      10      int    count      15.3±1ms
                DataFrame      10      int     skew     15.1±0.1ms
                DataFrame      10      int     kurt     15.5±0.3ms
                DataFrame      10      int     sum      13.3±0.6ms
                DataFrame      10     float   median     71.4±2ms
                DataFrame      10     float    mean     7.94±0.6ms
                DataFrame      10     float    max      8.96±0.3ms
                DataFrame      10     float    min      8.88±0.1ms
                DataFrame      10     float    std       12.0±3ms
                DataFrame      10     float   count      16.1±2ms
                DataFrame      10     float    skew     10.1±0.4ms
                DataFrame      10     float    kurt     10.4±0.7ms
                DataFrame      10     float    sum      8.19±0.6ms
                DataFrame     1000     int    median    147±0.9ms
                DataFrame     1000     int     mean      11.3±1ms
                DataFrame     1000     int     max      9.02±0.1ms
                DataFrame     1000     int     min     9.06±0.08ms
                DataFrame     1000     int     std      12.2±0.9ms
                DataFrame     1000     int    count      19.9±5ms
                DataFrame     1000     int     skew      12.7±1ms
                DataFrame     1000     int     kurt      9.19±3ms
                DataFrame     1000     int     sum       11.2±1ms
                DataFrame     1000    float   median     143±2ms
                DataFrame     1000    float    mean     3.99±0.6ms
                DataFrame     1000    float    max      8.90±0.1ms
                DataFrame     1000    float    min      9.01±0.1ms
                DataFrame     1000    float    std       11.9±3ms
                DataFrame     1000    float   count      19.8±5ms
                DataFrame     1000    float    skew     6.19±0.9ms
                DataFrame     1000    float    kurt     5.73±0.8ms
                DataFrame     1000    float    sum      3.61±0.7ms
                  Series       10      int    median     77.0±1ms
                  Series       10      int     mean     16.3±0.5ms
                  Series       10      int     max      10.3±0.2ms
                  Series       10      int     min      10.4±0.2ms
                  Series       10      int     std      19.2±0.3ms
                  Series       10      int    count      28.8±4ms
                  Series       10      int     skew     18.4±0.4ms
                  Series       10      int     kurt     18.0±0.4ms
                  Series       10      int     sum      16.7±0.2ms
                  Series       10     float   median    72.8±0.7ms
                  Series       10     float    mean     8.39±0.6ms
                  Series       10     float    max      10.4±0.1ms
                  Series       10     float    min      10.3±0.1ms
                  Series       10     float    std      15.3±0.1ms
                  Series       10     float   count      28.6±4ms
                  Series       10     float    skew      11.3±1ms
                  Series       10     float    kurt      11.9±1ms
                  Series       10     float    sum      8.23±0.6ms
                  Series      1000     int    median     151±2ms
                  Series      1000     int     mean     15.0±0.3ms
                  Series      1000     int     max      10.2±0.1ms
                  Series      1000     int     min      10.3±0.2ms
                  Series      1000     int     std      15.7±0.3ms
                  Series      1000     int    count      28.5±4ms
                  Series      1000     int     skew     16.2±0.3ms
                  Series      1000     int     kurt     16.5±0.2ms
                  Series      1000     int     sum      14.3±0.4ms
                  Series      1000    float   median     153±5ms
                  Series      1000    float    mean     14.8±0.3ms
                  Series      1000    float    max      10.2±0.1ms
                  Series      1000    float    min      10.4±0.2ms
                  Series      1000    float    std      17.3±0.3ms
                  Series      1000    float   count      28.6±4ms
                  Series      1000    float    skew     16.3±0.3ms
                  Series      1000    float    kurt     16.1±0.3ms
                  Series      1000    float    sum      14.4±0.2ms
              ============= ======== ======= ======== =============


BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```
</details>

File: pandas/tests/window/test_rolling.py (hunk 1 above)
Comment (AlexKirko) [reply pandas/tests/window/test_rolling.py]: Thanks, created a new issue. Will be happy to submit a PR in a day or two.

File: pandas/_libs/window/aggregations.pyx (hunk 4 above)
Comment (mroeschke) [reply pandas/_libs/window/aggregations.pyx:1076]: Great thanks!

**Review by mroeschke — APPROVED** [review]

**Comment by AlexKirko** [comment]
@jreback 
Between the new issue and the changes made, I think all the comments are covered. Is there anything else we should do before merging?

**Comment by jreback** [comment]
will looks again soon

**Comment by AlexKirko** [comment]
Changed the `ForwardIndexer` implementation to mirroring the one I will propose exposing to users in a separate PR.

**Comment by AlexKirko** [comment]
Some pipelines are being laggy. Should have nothing to do with this PR, but I restarted the tests just in case.

**Comment by AlexKirko** [comment]
Hm. py37_np_dev seems to be hanging for everyone

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @AlexKirko

**Comment by AlexKirko** [comment]
Thanks!

**Pull request merged**